src/ImageSharp/Formats/ImageDecoder.cs

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) Six Labors.
+// Licensed under the Six Labors Split License.
+
+using SixLabors.ImageSharp.IO;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.ImageSharp.Processing;
+
+namespace SixLabors.ImageSharp.Formats;
+
+/// <summary>
+/// Acts as a base class for image decoders.
+/// Types that inherit this decoder are required to implement cancellable synchronous decoding operations only.
+/// </summary>
+public abstract class ImageDecoder : IImageDecoder
+{
+    /// <inheritdoc/>
+    public Image<TPixel> Decode<TPixel>(DecoderOptions options, Stream stream)
+        where TPixel : unmanaged, IPixel<TPixel>
+        => WithSeekableStream(
```

Comment:
It's for when the decoder is called directly. Duplication is avoided because the decoder is always using a seekable stream when called via `Image.Load`

Reply:
For me this is a sign, that something is off with the design. Would be nice to figure out how to do it only once.

Reply:
It's really only a nice to have for our own decoders. I'm happy to remove it and leave the direct decoder usage pipeline throwing instead.

Reply:
I think I want to keep this. The stream is always seekable if loaded via the `Image.Load` middleware and our checks in the decoder will avoid a duplicate operation and call the action directly.

When someone calls `ImageDecoder.Decode()` or the async equivalent directly our checks will avoid unnecessary copying and also ensure that the input stream is cloned to a seekable stream when required.